packages/hash/api/src/graph/knowledge/system-types/account.fields.ts

```diff
@@ -0,0 +1,58 @@
+import { ImpureGraphFunction, PureGraphFunction } from "../..";
+import { RESTRICTED_SHORTNAMES } from "../../util";
+import { getUserByShortname } from "./user";
+import { getOrgByShortname } from "./org";
+
+// Validations for shortnames
+/**
+ * @todo revisit (simple) shortname validation to make use of data type
+ *   constraints.
+ *   https://app.asana.com/0/0/1202900021005257/f
+ */
+export const shortnameMinimumLength = 4;
+export const shortnameMaximumLength = 24;
+
+const ALLOWED_SHORTNAME_CHARS = /^[a-zA-Z0-9-_]+$/;
+
+export const shortnameContainsInvalidCharacter: PureGraphFunction<
+  { shortname: string },
+  boolean
+> = ({ shortname }) => {
+  return !!shortname.search(ALLOWED_SHORTNAME_CHARS);
+};
+
+export const shortnameIsRestricted: PureGraphFunction<
+  { shortname: string },
+  boolean
+> = ({ shortname }): boolean => {
+  return RESTRICTED_SHORTNAMES.includes(shortname);
+};
+
+export const shortnameIsTaken: ImpureGraphFunction<
+  { shortname: string },
+  Promise<boolean>
+> = async (ctx, params) => {
+  /**
+   * @todo currently these method calls create circular dependencies on the User
+   *   and Org model classes. We should revisit model class interaction to see
+   *   if we can get around this.
+   *   https://app.asana.com/0/1202805690238892/1202890446280565/f
+   */
```

Comment:
Hmm that's a good point, I forgot about this dependency cycle that is left unresolved. Tbh I'm not sure what we should do here. I don't want to move `getUserByShortname` and `getOrgByShortname` out of their files just to prevent a dependency cycle, as this would be inconsistent with how we decide other methods such as where `getUserByKratosIdentityId` is located.

Reply:
We potentially solve this similarly to how we previously did, by exporting all functions (in `user.ts`, `org.ts`, `entity.ts`, etc) from a single file (in `graph/knowledge/index.ts` perhaps), and then export both `shortnameIsTaken`, `getOrgByShortname ` and `getUserByShortname` from that file so they can be imported without creating a direct dependency cycle. I'd be in favour of addressing this in follow-up though.

cc @kachkaev who is working on removing dependency cycles throughout the codebase

Reply:
I've updated the todo in https://github.com/hashintel/hash/pull/1681/commits/389dfaaa1acb4dab8becac099a4625df7a8dd43c